chainntnfs/current_chain_view.go

```diff
@@ -0,0 +1,123 @@
+package chainntnfs
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/btcsuite/btcd/wire"
+)
+
+// CurrentChainStateView is an interface that allows the querying of the most
+// up-to-date blockchain state with low overhead. Valid implementations of this
+// interface must track the latest chain state.
+type CurrentChainStateView interface {
+	// Gets the most recent block height known to the view
+	BestHeight() uint32
+	// Gets the most recent block header known to the view
+	BestBlockHeader() *wire.BlockHeader
+}
+
+// CurrentChainStateTracker is a tiny subsystem that tracks the blockchain tip
+// and saves the most recent tip information in memory for querying. It is a
+// valid implementation of CurrentChainStateView and additionally includes
+// methods for starting and stopping the system.
+type CurrentChainStateTracker struct {
+	quit         chan struct{}
+	wg           sync.WaitGroup
+	notifier     *ChainNotifier
+	changeStream *BlockEpochEvent
+	current      atomic.Pointer[BlockEpoch]
+}
+
+// This gets the most recent block height known to the CurrentChainStateTracker.
+func (t *CurrentChainStateTracker) BestHeight() uint32 {
+	epoch := t.current.Load()
+	if epoch == nil {
+		return 0
+	}
+
+	return uint32(epoch.Height)
+}
+
+// This gets the most recent block header known to the CurrentChainStateTracker.
+func (t *CurrentChainStateTracker) BestBlockHeader() *wire.BlockHeader {
+	epoch := t.current.Load()
+	if epoch == nil {
+		return nil
+	}
+
+	return epoch.BlockHeader
+}
+
+// This method is a helper that subscribes to the underlying BlockEpochEvent
+// stream and updates the internal values to match the new BlockEpochs that
+// are discovered.
+//
+// MUST be run as a goroutine.
+func (t *CurrentChainStateTracker) updateLoop() {
+	defer t.wg.Done()
+	for {
+		select {
+		case epoch, ok := <-t.changeStream.Epochs:
+			if ok {
+				t.current.Store(epoch)
+			} else {
+				Log.Error(
+					"dead epoch stream in " +
+						"CurrentChainStateTracker",
+				)
+
+				return
+			}
+		case <-t.quit:
+			t.current.Store(nil)
+			return
+		}
+	}
+}
+
+// This method starts the CurrentChainStateTracker. It is an error to start it
+// if it is already started.
+func (t *CurrentChainStateTracker) Start() error {
+	if t.changeStream != nil {
+		return fmt.Errorf("CurrentChainStateTracker is already started")
+	}
+	changeStream, err := (*t.notifier).RegisterBlockEpochNtfn(nil)
+	if err != nil {
+		return err
+	}
+	t.changeStream = changeStream
+	t.quit = make(chan struct{})
```

Comment:
This allows the tracker to be started, stopped, then started again. If we place the channel construction in the constructor itself then it would complicate the logic of the `Start()` function or we would introduce the limitation that it cannot be started again once it is stopped. Perhaps this is OK.

Reply:
>  If we place the channel construction in the constructor itself then it would complicate the logic of the Start() function or we would introduce the limitation that it cannot be started again once it is stopped. Perhaps this is OK.

Gotcha, yeah usually most of these are only ever started once, as their lifetime tracks the lifetime of the peer connection. So I think it'd be safe to move all the initialization into the main constructor.

Reply:
The constructor is responsible for building the object. The start method is responsible for actually doing things and kicking off goroutines.